src/App.tsx

```diff
@@ -1,24 +1,240 @@
+    }
+  }, [todos]);
+
+  const reverteCompletedTodo = useCallback((todosForRevert: Todo[]) => {
+    const todosForUpdate = todosForRevert;
+    const uncompletedTodosForUpdate = TodoFilter
+      .uncompletedTodos(todosForRevert);
+
+    const todosForUpdateId = uncompletedTodosForUpdate.length !== 0
+      ? uncompletedTodosForUpdate.map(todo => todo.id)
+      : todosForUpdate.map(todo => todo.id);
+
+    setUpdatingTodos(todosForUpdateId);
+
+    if (uncompletedTodosForUpdate.length !== 0) {
+      uncompletedTodosForUpdate.map(todo => changeTodoCompleted(todo.id));
+    } else {
+      todosForUpdate.map(todo => changeTodoCompleted(todo.id));
+    }
+  }, [todos]);
+
+  const updateTodo = useCallback((todoId: number | null, title: string) => {
+    const todoForUpdate = findTodoById(todos, todoId);
+
+    setLoading(true);
+
+    if (todoForUpdate !== null) {
+      todoForUpdate.title = title;
+
+      todoService.updateTodo(todoForUpdate)
+        .then(todo => {
+          setTodos(currentTodos => {
+            const newTodos = [...currentTodos];
+            const index = newTodos.findIndex(newTodo => {
+              return newTodo.id === todoForUpdate.id;
+            });
+
+            newTodos.splice(index, 1, todo);
+
+            return newTodos;
+          });
+        })
+        .catch((error) => {
+          setNotification(NOTIFICATION.UPDATE);
+          throw error;
+        })
+        .finally(() => {
+          setLoading(false);
+        });
+    }
+  }, [todos]);
+
+  const filteredTodos: Todo[] = useMemo(() => {
+    return TodoFilter.getFilteredTodos(todos, filterType);
+  }, [todos, filterType]);
+
   if (!USER_ID) {
     return <UserWarning />;
   }
 
   return (
-    <section className="section container">
-      <p className="title is-4">
-        Copy all you need from the prev task:
-        <br />
-        <a href="https://github.com/mate-academy/react_todo-app-add-and-delete#react-todo-app-add-and-delete">React Todo App - Add and Delete</a>
-      </p>
-
-      <p className="subtitle">Styles are already copied</p>
-    </section>
+    <div className="todoapp">
+      <h1 className="todoapp__title">todos</h1>
+
+      <div className="todoapp__content">
+        <header className="todoapp__header">
+          <Form
+            loading={loading}
+            todos={todos}
+            addTodo={(newTodo) => addTodo(newTodo)}
+            reverteCompletedTodo={
+              (todosForRevert) => reverteCompletedTodo(todosForRevert)
+            }
+            titleFocus={titleFocus}
+          />
+        </header>
+
+        <TodoList
+          listOfTodoId={listOfTodoId}
+          loading={loading}
+          todos={filteredTodos}
+          deleteTodo={deleteTodo}
+          tempTodo={tempTodo}
+          updateTodo={(todo, title) => updateTodo(todo, title)}
+          onChange={(todoId) => changeTodoCompleted(todoId)}
+          updatingTodos={updatingTodos}
+        />
+
+        {todos.length !== 0
```

Comment:
```suggestion
        {todos.length
```

Reply:
fixed

Reply:
In the code, I can still see `todos.length !== 0`